confirmation/migrations/0009_confirmation_expiry_date_backfill.py

```diff
@@ -0,0 +1,71 @@
+# Generated by Django 3.1.7 on 2021-03-31 20:47
+
+import time
+from datetime import timedelta
+
+from django.conf import settings
+from django.db import migrations, transaction
+from django.db.backends.postgresql.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+
+def set_expiry_date_for_existing_confirmations(
+    apps: StateApps, schema_editor: DatabaseSchemaEditor
+) -> None:
+    Confirmation = apps.get_model("confirmation", "Confirmation")
+    if not Confirmation.objects.exists():
+        return
+
+    # The values at the time of this migration
+    INVITATION = 2
+    UNSUBSCRIBE = 4
+    MULTIUSE_INVITE = 6
+
+    @transaction.atomic()
+    def backfill_confirmations_between(lower_bound: int, upper_bound: int) -> None:
+        confirmations = Confirmation.objects.filter(id__gte=lower_bound, id__lte=upper_bound)
+        for confirmation in confirmations:
+            if confirmation.type in (INVITATION, MULTIUSE_INVITE):
+                confirmation.expiry_date = confirmation.date_sent + timedelta(
+                    settings.INVITATION_LINK_VALIDITY_DAYS
+                )
+            elif confirmation.type == UNSUBSCRIBE:
+                confirmation.expiry_date = confirmation.date_sent + timedelta(1000000)
+            else:
+                confirmation.expiry_date = confirmation.date_sent + timedelta(
+                    settings.CONFIRMATION_LINK_DEFAULT_VALIDITY_DAYS
+                )
+        Confirmation.objects.bulk_update(confirmations, ["expiry_date"])
+
+    BATCH_SIZE = 1000
+
+    first_id = Confirmation.objects.earliest("id").id
+    last_id = Confirmation.objects.latest("id").id
+
+    id_range_lower_bound = first_id
+    id_range_upper_bound = first_id + BATCH_SIZE
+    while id_range_upper_bound <= last_id:
+        backfill_confirmations_between(id_range_lower_bound, id_range_upper_bound)
+        id_range_lower_bound = id_range_upper_bound + 1
+        id_range_upper_bound = id_range_lower_bound + BATCH_SIZE
+        time.sleep(0.1)
+
+    if last_id > id_range_lower_bound:
+        # Backfill for the last batch.
+        backfill_confirmations_between(id_range_lower_bound, last_id)
```

Comment:
I applied this diff to simplify the migration and make it a bit easier to reason about:

```
diff --git a/confirmation/migrations/0009_confirmation_expiry_date_backfill.py b/confirmation/migrations/0009_confirmation_expiry_date_backfill.py
index b9c93bd1ff..bc7a74b9df 100644
--- a/confirmation/migrations/0009_confirmation_expiry_date_backfill.py
+++ b/confirmation/migrations/0009_confirmation_expiry_date_backfill.py
@@ -27,13 +27,14 @@ def set_expiry_date_for_existing_confirmations(
         for confirmation in confirmations:
             if confirmation.type in (INVITATION, MULTIUSE_INVITE):
                 confirmation.expiry_date = confirmation.date_sent + timedelta(
-                    settings.INVITATION_LINK_VALIDITY_DAYS
+                    days=settings.INVITATION_LINK_VALIDITY_DAYS
                 )
             elif confirmation.type == UNSUBSCRIBE:
-                confirmation.expiry_date = confirmation.date_sent + timedelta(1000000)
+                # Unsubscribe links never expire, which we apparently implement as in 1M days.
+                confirmation.expiry_date = confirmation.date_sent + timedelta(days=1000000)
             else:
                 confirmation.expiry_date = confirmation.date_sent + timedelta(
-                    settings.CONFIRMATION_LINK_DEFAULT_VALIDITY_DAYS
+                    days=settings.CONFIRMATION_LINK_DEFAULT_VALIDITY_DAYS
                 )
         Confirmation.objects.bulk_update(confirmations, ["expiry_date"])
 
@@ -44,16 +45,12 @@ def set_expiry_date_for_existing_confirmations(
 
     id_range_lower_bound = first_id
     id_range_upper_bound = first_id + BATCH_SIZE
-    while id_range_upper_bound <= last_id:
+    while id_range_lower_bound <= last_id:
         backfill_confirmations_between(id_range_lower_bound, id_range_upper_bound)
         id_range_lower_bound = id_range_upper_bound + 1
         id_range_upper_bound = id_range_lower_bound + BATCH_SIZE
         time.sleep(0.1)
 
-    if last_id > id_range_lower_bound:
-        # Backfill for the last batch.
-        backfill_confirmations_between(id_range_lower_bound, last_id)
-
 
 class Migration(migrations.Migration):
     atomic = False
```